packages/marshal/src/types.js

```diff
@@ -2,70 +2,317 @@
 /// <reference path="extra-types.d.ts" />
 
 /**
- * @typedef { "bigint" | "boolean" | "null" | "number" | "string" | "symbol" | "undefined" | "copyArray" | "copyRecord" | "copyError" | "promise" | "presence" } PassStyle
- * TODO "presence" above should indirect through REMOTE_STYLE to prepare
- * for changing it to "remotable"
+ * @typedef { "undefined" | "null" |
+ *   "boolean" | "number" | "bigint" | "string" | "symbol" |
+ *   "copyArray" | "copyRecord" | "copyError" | "remotable" | "promise"
+ * } PassStyle
+ *
+ * TODO: We need to add a binary blob type to the PassStyles above and the
+ * LeafData below, consisting of an uninterpreted sequence of octets (8 bit
+ * bytes). The ideal JS representation would be a hardened `Uint8Array` except
+ * that Uint8Arrays cannot be frozen. Instead the likely JS representation
+ * will be a hardened ArrayBuffer.
  */
 
-// TODO declare more precise types throughout this file, so the type system
-// and IDE can be more helpful.
+/**
+ * @typedef { void | null | boolean | number | bigint | string | symbol
+ * } LeafData
+ * Corresponding PassStyle values are
+ * "undefined", "null", "boolean", "number", "bigint", "string", "symbol".
+ * A LeafData is a PassableKey and can be compared for equivalence with
+ * `sameKey`.
+ * LeafData can be sorted according to a yet-to-be-defined full order over all
+ * LeafData.
+ *    * We inherit JS's peculiar co-existence of two bottom values,
+ *      `null` and `undefined`.
+ *    * Following JS, `number` is an IEEE double precision floating point
+ *      value which may be `NaN`, `Infinity`, or `-Infinity`.
+ *    * There is only one NaN. A binary encoding therefore needs to
+ *      canonicalize the NaN. The one NaN compares equivalent to itself.
+ *    * There is only one 0, the IEEE non-negative zero. Negative
+ *      zero must be canonicalized to non-negative zero, which must alway
+ *      decode as non-negative zero. The one zero compares as equivalent to
+ *      itself.
+ *    * strings include all Unicode strings representable in JSON or
+ *      utf8, even if they contain unpaired UTF-16 surrogates.
+ *    * The `symbol` category includes only symbols which can be recreated
+ *      from string data. It therefore excludes JS unregistered symbols or
+ *      lisp/scheme gensyms. From JS it currently includes only
+ *      `Symbol.asyncIterator` (a so-called well-known symbol). From JS
+ *      it can include at most the well known symbols and the registered
+ *      symbols. (Because symbol registration names can collide with
+ *      well known symbol names, we will need another hilbert hotel
+ *      encoding to allow well known symbols and all registered symbols.)
+ *      From lisp, it can include symbols made from interned strings.
+ */
 
 /**
- * @typedef {*} Passable
+ * @typedef {*} PrimaryRemotable
```

Comment:
A `PrimaryRemotable` should be something more specialised like:
```suggestion
 * @typedef {Record<string | symbol, Function>} PrimaryRemotable
```